py/desispec/workflow/schedule.py

```diff
@@ -0,0 +1,127 @@
+            destrank = worker * self.group_size + grouprank + 1
+            self.job_buff[0] = job
+            self.comm.Send(self.job_buff,dest=destrank)
+            reqs.append(self.comm.Irecv(self.job_buff,source=destrank))
+        return reqs
+
+    def _checkreqlist(self,reqs):
+
+        # check if all processes with handles in reqs have reported back
+        for req in reqs:
+            if not req.Test():
+                return False
+        return True
+
+    def _schedule(self):
+
+        # bookkeeping
+        waitlist        = [] # message handles for pending worker groups
+        worker_groups   = [] # worker assigned
+
+        # start by assigning Ngroups jobs to each of the Ngroup groups
+        nextjob=0
+        for job in range(self.Ngroups):
+            worker = nextjob
+            reqs=self._assign_job(worker,nextjob)
+            waitlist.append(reqs)
+            worker_groups.append(worker)
+            nextjob += 1
+
+        # the scheduler waits for jobs to be completed;
+        # when one is complete it assigns the next job
+        # until there are none left
+        Ncomplete = 0
+        while(Ncomplete < self.Njobs):
+            # iterate over list of currently pending group of processes
+            for i in range(len(waitlist)):
+                # check for completion of all processes in this worker group
+                if self._checkreqlist(waitlist[i]):
+                    # all ranks group doing job corresponding to place i in waitlist
+                    # have returned; identify this worker group and remove it from the
+                    # waitlist and worker list
+                    worker = worker_groups[i]
+                    Ncomplete += 1
+                    waitlist.pop(i)
+                    worker_groups.pop(i)
+                    if nextjob < self.Njobs:
+                        # more jobs to do; assign processes in group worker
+                        # the job with index nextjob, increment
+                        reqs=self._assign_job(worker,nextjob)
+                        waitlist.append(reqs)
+                        worker_groups.append(worker)
+                        nextjob += 1
+                    # waitlist has been modified so exit waitlist loop with break
+                    break
+
+        # no more jobs to assign; dismiss all processes in all groups by 
+        # assigning job = -1, causing all workers processes to return
+        for worker in range(self.Ngroups): self._assign_job(worker,-1)
+
+        return
+
+    def _work(self):
+
+        # listen for job assignments from the scheduler
+        while True:
+            self.comm.Recv(self.job_buff,source=0) # receive assignment from rank=0 scheduler
+            job = self.job_buff[0]                 # unpack job index
+            if job < 0: return                     # job < 0 means no more jobs to do
+            self._workfunc(self.groupcomm,job)     # call work function for job
```

Comment:
I think this should be wrapped in a generic try/except block so that if the function raises an exception, it doesn't end up blocking the scheduler until the job times out (since that worker will never get to the final barrier).  There might be other solutions, but please test the scheduler with the specific case of some jobs raising exceptions and others succeeding, and making sure that doesn't lead to a hang.